src/libraries/System.Data.Common/src/System/Data/DataColumn.cs

```diff
@@ -1525,6 +1540,8 @@ internal void Copy(int srcRecordNo, int dstRecordNo)
 
         // Prevent inlining so that reflection calls are not moved to caller that may be in a different assembly that may have a different grant set.
         [MethodImpl(MethodImplOptions.NoInlining)]
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "Only using parameterless constructor which is safe. Class has appropriate attribute.")]
```

Comment:
Here are the warnings I'm getting without this:

```
d:\src\runtime-2\src\libraries\System.Data.Common\src\System\Data\DataColumn.cs(1545,13): Trim analysis warning IL2026: System.Data.DataColumn.Clone(): Using method 'System.Data.DataColumn.DataColumn(String,Type,String,MappingType)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Members from serialized types or types used in expressions may be trimmed if not referenced directly.. [d:\src\runtime-2\src\libraries\src.proj]
d:\src\runtime-2\src\libraries\System.Data.Common\src\System\Data\DataColumn.cs(1545,13): Trim analysis warning IL2026: System.Data.DataColumn.Clone(): Using method 'System.Data.DataColumn.DataColumn(String,Type,String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Members from serialized types or types used in expressions may be trimmed if not referenced directly.. [d:\src\runtime-2\src\libraries\src.proj]
```

So basically I'm getting warning about constructor which is not being called in here... thoughts?

Reply:
The warnings come from the `GetType()` call. The `DataColumn` type is annotated with `DynamicallyAccessedMemberTypes.PublicConstructors` and so that call will end up marking all of the public constructors.

To fix this, I think the `DataColumn` should be marked only with `PublicParameterlessConstructor` since that's what it needs in the `Clone` method. I tried this locally and changing that annotation makes the suppression on the Clone unnecessary. There are no warnings that way.

Reply:
> There are no warnings that way.

This scenario is a tricky one to understand. It took me a few times of seeing it before I fully understood it.

Since the `DataColumn` class is marked with `PublicConstructors`, calls to `GetType()` see that the class is marked to preserve "all public constructors". So the trimmer looks through the public constructors, and sees that 2 of them are marked with `[RequiresUnreferencedCode]`. Since you said "preserve all constructors" and some of the constructors are "unsafe", the trimmer is warning that you may be reflecting on unsafe methods, even though you actually aren't here. It is "two levels" of attributes that you have to combine together to understand the scenario.

Changing the `[DAM]` on `DataColumn` to `PublicParameterlessConstructor` tells the trimmer to preserve just the parameterless constructor. And since that one isn't marked `[RUC]`, there is no warning.